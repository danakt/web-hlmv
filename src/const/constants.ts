/** Supported model format version */
export const VERSION = 10;

/** Maximum number of bone controllers per bone */
export const MAX_PER_BONE_CONTROLLERS = 6;

/** Flag of texture masking */
export const NF_MASKED = 0x0040;

/** Number of colors */
export const PALETTE_ENTRIES = 256;

/** Number of channels for RGB color. Was "PALETTE_CHANNELS" */
export const RGB_SIZE = 3;

/** Number of channels for RGBA color. Was "PALETTE_CHANNELS_ALPHA" */
export const RGBA_SIZE = 4;

/** Total size of a palette, in bytes. */
export const PALETTE_SIZE = PALETTE_ENTRIES * RGB_SIZE;

/** The index in a palette where the alpha color is stored. Used for transparent textures. */
export const PALETTE_ALPHA_INDEX = 255 * RGB_SIZE;

/** Number of bones allowed at source movement */
export const MAX_SRCBONES = 512;

/** Number of axles in 3d space */
export const AXLES_NUM = 3;

/** Animation value items index constants */
export const enum ANIM_VALUE {
  VALUE = 0,
  VALID,
  TOTAL
}

/** Triangle fan type */
export const TRIANGLE_FAN = 0;

/** Triangle strip type */
export const TRIANGLE_STRIP = 1;

/** Motion flag X */
export const MOTION_X = 0x0001;

/** Motion flag Y */
export const MOTION_Y = 0x0002;

/** Motion flag Z */
export const MOTION_Z = 0x0004;

/** Controller that wraps shortest distance */
export const RLOOP = 0x8000;

/** Default interface background color */
export const INITIAL_UI_BACKGROUND = '#4d7f7e';
