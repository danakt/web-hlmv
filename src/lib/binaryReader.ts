// eslint-disable-next-line no-unused-vars
import { Struct, StructResult } from './dataTypes'; // FIXME

/**
 * Parses binary buffer (wrapped to DataView) according to a structure
 * description
 * @param dataView The DataView object
 * @param struct Structure description. The structure can't contain keys
 * starting with a digit due to the peculiarities of the javascript engine.
 * Otherwise, the reading result may be corrupted.
 * @param byteOffset Offset in buffer to read, "0" by default
 * @returns The structure applying result
 */
export const readStruct = function<S extends Struct>(
  dataView: DataView,
  struct: S,
  byteOffset: number = 0
): StructResult<S> {
  let offset: number = byteOffset;
  const structResult = {} as StructResult<S>;

  for (const key in struct) {
    (structResult as any)[key] = struct[key].getValue(dataView, offset);

    offset += struct[key].byteLength;
  }

  return structResult;
};

/**
 * @todo Describe me
 *
 * @param dataView The DataView object
 * @param struct Structure description. The structure can't contain keys
 * starting with a digit due to the peculiarities of the javascript engine.
 * Otherwise, the reading result may be corrupted.
 * @param byteOffset Offset in buffer to read, "0" by default
 * @param times Times of structure repeating
 * @returns The array of structure applying result
 */
export const readStructMultiple = function<S extends Struct>(
  dataView: DataView,
  struct: S,
  byteOffset: number = 0,
  times: number = 1
): StructResult<S>[] {
  let offset: number = byteOffset;
  const result: StructResult<S>[] = [];

  for (let i = 0; i < times; i++) {
    const structResult = {} as StructResult<S>;

    for (const key in struct) {
      (structResult as any)[key] = struct[key].getValue(dataView, offset);

      offset += struct[key].byteLength;
    }

    result[i] = structResult;
  }

  return result;
};

/**
 * Returns length of a structure
 * @param struct Structure description
 */
export const getStructLength = <S extends Struct>(struct: S): number => {
  let length: number = 0;

  for (const key in struct) {
    length += struct[key].byteLength;
  }

  return length;
};
