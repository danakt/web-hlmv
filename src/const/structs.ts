import { StructResult, int, float, string, array, vec3, ushort } from '../lib/dataTypes';
import { MAX_PER_BONE_CONTROLLERS } from './constants';

/**
 * Head of mdl-file
 */
export const header = {
  /** Model format ID */
  id: int,
  /** Format version number */
  version: int,
  /** The internal name of the model */
  name: string(64),
  /** Data size of MDL file in bytes */
  length: int,
  /** Position of player viewpoint relative to model origin */
  eyePosition: vec3,
  /** Corner of model hull box with the least X/Y/Z values */
  max: vec3,
  /** Opposite corner of model hull box */
  min: vec3,
  /** Min position of view bounding box */
  bbmin: vec3,
  /** Max position of view bounding box */
  bbmax: vec3,
  /**
   * Binary flags in little-endian order.
   * ex (00000001, 00000000, 00000000, 11000000) means flags for position
   * 0, 30, and 31 are set. Set model flags section for more information
   */
  flags: int,

  // After this point, the header contains many references to offsets
  // within the MDL file and the number of items at those offsets.
  // Offsets are from the very beginning of the file.
  // Note that indexes/counts are not always paired and ordered consistently.

  /** Number of bones */
  numBones: int,
  /** Offset of first data section */
  boneIndex: int,
  /** Number of bone controllers */
  numBoneControllers: int,
  /** Offset of bone controllers */
  boneControllerIndex: int,
  /** Number of complex bounding boxes */
  numHitboxes: int,
  /** Offset of hit boxes */
  hitBoxIndex: int,
  /** Number of sequences */
  numSeq: int,
  /** Offset of sequences */
  seqIndex: int,
  /** Number of demand loaded sequences */
  numSeqGroups: int,
  /** Offset of demand loaded sequences */
  seqGroupIndex: int,
  /** Number of raw textures */
  numTextures: int,
  /** Offset of raw textures */
  textureIndex: int,
  /** Offset of textures data */
  textureDataIndex: int,
  /** Number of replaceable textures */
  numSkinRef: int,
  numSkinFamilies: int,
  skinIndex: int,
  /** Number of body parts */
  numBodyParts: int,
  /** Index of body parts */
  bodyPartIndex: int,
  /** Number queryable attachable points */
  numAttachments: int,
  attachmentIndex: int,
  // This seems to be obsolete.
  // Probably replaced by events that reference external sounds?
  soundTable: int,
  soundIndex: int,
  soundGroups: int,
  soundGroupIndex: int,
  /** Animation node to animation node transition graph */
  numTransitions: int,
  transitionIndex: int
};

export type Header = StructResult<typeof header>;

/**
 * Bone description
 */
export const bone = {
  /** Bone name for symbolic links */
  name: string(32),
  /** Parent bone */
  parent: int,
  /** ?? */
  flags: int,
  /** Bone controller index, -1 == none */
  boneController: array(MAX_PER_BONE_CONTROLLERS, int),
  /** Default DoF values */
  value: array(MAX_PER_BONE_CONTROLLERS, float),
  /** Scale for delta DoF values */
  scale: array(MAX_PER_BONE_CONTROLLERS, float)
};

export type Bone = StructResult<typeof bone>;

/**
 * Bone controllers
 */
export const boneController = {
  bone: int,
  type: int,
  start: float,
  end: float,
  rest: int,
  index: int
};

export type BoneController = StructResult<typeof boneController>;

/**
 * Attachment
 */
export const attachment = {
  name: string(32),
  type: int,
  bone: int,
  /** Attachment point */
  org: vec3,
  vectors: array(3, vec3)
};

export type Attachment = StructResult<typeof attachment>;

/**
 * Bounding boxes
 */
export const boundingBox = {
  bone: int,
  /** Intersection group */
  group: int,
  /** Bounding box */
  bbmin: vec3,
  bbmax: vec3
};

export type BoundingBox = StructResult<typeof boundingBox>;

/**
 * Sequence description
 */
export const seqDesc = {
  /** Sequence label */
  label: string(32),

  /** Frames per second */
  fps: float,
  /** Looping/non-looping flags */
  flags: int,

  activity: int,
  actWeight: int,

  numEvents: int,
  eventIndex: int,

  /** Number of frames per sequence */
  numFrames: int,

  /** Number of foot pivots */
  numPivots: int,
  pivotIndex: int,

  motionType: int,
  motionBone: int,
  linearMovement: vec3,
  autoMovePosIndex: int,
  autoMoveAngleIndex: int,

  /** Per sequence bounding box */
  bbmin: vec3,
  bbmax: vec3,

  numBlends: int,
  /** "anim" pointer relative to start of sequence group data */
  animIndex: int,

  // [blend][bone][X, Y, Z, XR, YR, ZR]
  /** X, Y, Z, XR, YR, ZR */
  blendType: array(2, int),
  /** Starting value */
  blendStart: array(2, float),
  /** Ending value */
  blendEnd: array(2, float),
  blendParent: int,

  /** Sequence group for demand loading */
  seqGroup: int,

  /** Transition node at entry */
  entryNode: int,
  /** Transition node at exit */
  exitNode: int,
  /** Transition rules */
  nodeFlags: int,

  /** Auto advancing sequences */
  nextSeq: int
};

export type SequenceDesc = StructResult<typeof seqDesc>;

/**
 * Demand loaded sequence groups
 */
export const seqGroup = {
  /** Textual name */
  label: string(32),
  /** File name */
  name: string(64),
  /** Was "cache" - index pointer */
  unused1: int,
  /** Was "data" - hack for group 0 */
  unused2: int
};

export type SequenceGroup = StructResult<typeof seqGroup>;

/**
 * Body part index
 */
export const bodyPart = {
  name: string(64),
  numModels: int,
  base: int,
  /** Index into models array */
  modelIndex: int
};

export type BodyPart = StructResult<typeof bodyPart>;

/**
 * Texture info
 */
export const texture = {
  /** Texture name */
  name: string(64),
  /** Flags */
  flags: int,
  /** Texture width */
  width: int,
  /** Texture height */
  height: int,
  /** Texture data offset */
  index: int
};

export type Texture = StructResult<typeof texture>;

/**
 * Sub models
 */
export const subModel = {
  name: string(64),

  type: int,

  boundingRadius: float,

  numMesh: int,
  meshIndex: int,

  /** Number of unique vertices */
  numVerts: int,
  /** Vertex bone info */
  vertInfoIndex: int,
  /** Vertex vec3 */
  vertIndex: int,
  /** Number of unique surface normals */
  numNorms: int,
  /** Normal bone info */
  normInfoIndex: int,
  /** Normal vec3 */
  normIndex: int,

  /** Deformation groups */
  numGroups: int,
  groupIndex: int
};

export type SubModel = StructResult<typeof subModel>;

/**
 * Mesh info
 */
export const mesh = {
  numTris: int,
  triIndex: int,
  skinRef: int,
  /** Per mesh normals */
  numNorms: int,
  /** Normal vec3_t */
  normIndex: int
};

export type Mesh = StructResult<typeof mesh>;

/**
 * Animation description
 */
export const animation = {
  offset: array(6, ushort)
};

export type Animation = StructResult<typeof animation>;
