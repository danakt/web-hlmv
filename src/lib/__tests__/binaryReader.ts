import * as fs from 'fs';
import * as path from 'path';
import FastDataView from 'fast-dataview';
import * as structs from '../../const/structs';
import { readStruct, readStructMultiple } from '../binaryReader';
import {
  int8,
  byte,
  uint8,
  ubyte,
  short,
  int16,
  uint16,
  ushort,
  int32,
  int,
  uint32,
  uint,
  float32,
  float,
  float64,
  double,
  string,
  array,
  vec3,
  skip
} from '../dataTypes';

// Loading model for testing
const leetPath = path.resolve(__dirname, '../../__mock__/leet.mdl');
const leetBuffer: ArrayBuffer = fs.readFileSync(leetPath).buffer;

test('should parse some struct in binary file', () => {
  const dataView = new FastDataView(leetBuffer);
  const header = readStruct(dataView, structs.header);
  expect(header).toMatchSnapshot('leet header');
});

test('should parse some struct in binary file with offset', () => {
  const dataView = new FastDataView(leetBuffer);
  const header = readStruct(dataView, structs.header);
  const hitBox = readStruct(dataView, structs.boundingBox, header.hitBoxIndex);
  expect(hitBox).toMatchSnapshot('some hitbox');
});

test('should parse multiple structs in binary file', () => {
  const dataView = new FastDataView(leetBuffer);
  const header = readStruct(dataView, structs.header);
  const texturesInfo = readStructMultiple(dataView, structs.texture, header.textureIndex, header.numTextures);
  expect(texturesInfo).toMatchSnapshot('leet textures info');
});

test('should parse multiple structs in binary file with offset', () => {
  const dataView = new FastDataView(leetBuffer);
  const header = readStructMultiple(dataView, structs.header)[0];
  expect(header).toMatchSnapshot('leet header');
});

test('should get byte from buffer', () => {
  const dataView = new DataView(new Int8Array([-1]).buffer);
  expect(int8.getValue(dataView, 0)).toBe(-1);
  expect(byte.getValue(dataView, 0)).toBe(-1);
});

test('should get unsigned byte from buffer', () => {
  const dataView = new DataView(new Uint8Array([-1]).buffer);
  expect(uint8.getValue(dataView, 0)).toBe(255);
  expect(ubyte.getValue(dataView, 0)).toBe(255);
});

test('should get short integer from buffer', () => {
  const dataView = new DataView(new Int16Array([-1]).buffer);
  expect(int16.getValue(dataView, 0)).toBe(-1);
  expect(short.getValue(dataView, 0)).toBe(-1);
});

test('should get unsigned short integer from buffer', () => {
  const dataView = new DataView(new Uint16Array([-1]).buffer);
  expect(uint16.getValue(dataView, 0)).toBe(65535);
  expect(ushort.getValue(dataView, 0)).toBe(65535);
});

test('should get integer from buffer', () => {
  const dataView = new DataView(new Int32Array([-1]).buffer);
  expect(int32.getValue(dataView, 0)).toBe(-1);
  expect(int.getValue(dataView, 0)).toBe(-1);
});

test('should get unsigned integer from buffer', () => {
  const dataView = new DataView(new Uint32Array([-1]).buffer);
  expect(uint32.getValue(dataView, 0)).toBe(4294967295);
  expect(uint.getValue(dataView, 0)).toBe(4294967295);
});

test('should get float from buffer', () => {
  const dataView = new DataView(new Float32Array([-Math.PI]).buffer);
  // Inaccurate value of PI
  expect(float32.getValue(dataView, 0)).toBe(-3.1415927410125732);
  expect(float.getValue(dataView, 0)).toBe(-3.1415927410125732);
});

test('should get double float from buffer', () => {
  const dataView = new DataView(new Float64Array([-Math.PI]).buffer);
  expect(float64.getValue(dataView, 0)).toBe(-Math.PI);
  expect(double.getValue(dataView, 0)).toBe(-Math.PI);
});

test('should get string from buffer', () => {
  const dataView = new DataView(new Int8Array([104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]).buffer);
  expect(string(11).getValue(dataView, 0)).toBe('hello world');
});

test('should get array of bytes from buffer', () => {
  const dataView = new DataView(new Int8Array([1, 2, 3]).buffer);
  expect(array(3, byte).getValue(dataView, 0)).toEqual(new Int8Array([1, 2, 3]));
});

test('should get vector 3 from buffer', () => {
  const dataView = new DataView(new Float32Array([0.5, 0.25, 0.125]).buffer);
  expect(vec3.getValue(dataView, 0)).toEqual(new Float32Array([0.5, 0.25, 0.125]));
});

test('should create skipping data type', () => {
  expect(skip(8).byteLength).toEqual(8);
  expect(skip(int).byteLength).toEqual(4);
});
