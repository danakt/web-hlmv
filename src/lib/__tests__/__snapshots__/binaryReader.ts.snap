// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse multiple structs in binary file with offset: leet header 1`] = `
Object {
  "attachmentIndex": 6428,
  "bbmax": Float32Array [
    0,
    0,
    0,
  ],
  "bbmin": Float32Array [
    0,
    0,
    0,
  ],
  "bodyPartIndex": 2094732,
  "boneControllerIndex": 6404,
  "boneIndex": 244,
  "eyePosition": Float32Array [
    0,
    0,
    0,
  ],
  "flags": 0,
  "hitBoxIndex": 6604,
  "id": 1414743113,
  "length": 2401992,
  "max": Float32Array [
    0,
    0,
    0,
  ],
  "min": Float32Array [
    0,
    0,
    0,
  ],
  "name": "leet\\\\leet.mdl",
  "numAttachments": 2,
  "numBodyParts": 2,
  "numBoneControllers": 1,
  "numBones": 55,
  "numHitboxes": 21,
  "numSeq": 111,
  "numSeqGroups": 1,
  "numSkinFamilies": 1,
  "numSkinRef": 3,
  "numTextures": 3,
  "numTransitions": 0,
  "seqGroupIndex": 2094628,
  "seqIndex": 2072508,
  "skinIndex": 2117056,
  "soundGroupIndex": 0,
  "soundGroups": 0,
  "soundIndex": 0,
  "soundTable": 0,
  "textureDataIndex": 2117064,
  "textureIndex": 2116816,
  "transitionIndex": 2094732,
  "version": 10,
}
`;

exports[`should parse multiple structs in binary file: leet textures info 1`] = `
Array [
  Object {
    "flags": 0,
    "height": 512,
    "index": 2117064,
    "name": "Arab_dmbase1.bmp",
    "width": 512,
  },
  Object {
    "flags": 3,
    "height": 64,
    "index": 2379976,
    "name": "Chrome3.bmp",
    "width": 64,
  },
  Object {
    "flags": 0,
    "height": 128,
    "index": 2384840,
    "name": "Backpack1.BMP",
    "width": 128,
  },
]
`;

exports[`should parse some struct in binary file with offset: some hitbox 1`] = `
Object {
  "bbmax": Float32Array [
    4,
    5.559999942779541,
    6.75,
  ],
  "bbmin": Float32Array [
    -4.690000057220459,
    -4.440000057220459,
    -6.75,
  ],
  "bone": 1,
  "group": 3,
}
`;

exports[`should parse some struct in binary file: leet header 1`] = `
Object {
  "attachmentIndex": 6428,
  "bbmax": Float32Array [
    0,
    0,
    0,
  ],
  "bbmin": Float32Array [
    0,
    0,
    0,
  ],
  "bodyPartIndex": 2094732,
  "boneControllerIndex": 6404,
  "boneIndex": 244,
  "eyePosition": Float32Array [
    0,
    0,
    0,
  ],
  "flags": 0,
  "hitBoxIndex": 6604,
  "id": 1414743113,
  "length": 2401992,
  "max": Float32Array [
    0,
    0,
    0,
  ],
  "min": Float32Array [
    0,
    0,
    0,
  ],
  "name": "leet\\\\leet.mdl",
  "numAttachments": 2,
  "numBodyParts": 2,
  "numBoneControllers": 1,
  "numBones": 55,
  "numHitboxes": 21,
  "numSeq": 111,
  "numSeqGroups": 1,
  "numSkinFamilies": 1,
  "numSkinRef": 3,
  "numTextures": 3,
  "numTransitions": 0,
  "seqGroupIndex": 2094628,
  "seqIndex": 2072508,
  "skinIndex": 2117056,
  "soundGroupIndex": 0,
  "soundGroups": 0,
  "soundIndex": 0,
  "soundTable": 0,
  "textureDataIndex": 2117064,
  "textureIndex": 2116816,
  "transitionIndex": 2094732,
  "version": 10,
}
`;
